### PR TITLE
Use the UpdateKernel for config:import.

### LINKED ISSUE #3574: Drush becomes unusable if a command requires a new module
If a drush command depends on a service in a module that is not yet enabled, drush can become unusable.

For example, let's say `Custom Module A` adds a Drush command and in it's `drush.services.yml` file uses a service exposed by `Contrib Module B`. In a minor release, `Contrib Module B` alters it's service definition to use a class from a newly added dependency. Note it's not necessarily breaking API contracts; perhaps a new module has some nice implementation of a core-provided interface.

The problem arises when trying to use Drush to run update hooks. As a developer or site builder, I expect I can run `composer update && drush -y updatedb`. However, while discovering all implemented Drush commands, building the container fails because the newly-required class exists in a module that is not yet enabled. As an end user, the only workaround is to use the user interface to run update hooks, or to do two deployments.

There is a way around this [by using a service provider instead of the `drush.services.yml` file](https://github.com/Lullabot/media_mpx/pull/52/files). That lets code do basic checks like `class_exists(...)` before defining a service. However, the services are not as discoverable as a YAML file so the DX for new developers isn't great. As is, I think this is required if you depend on any services or classes outside of the command's module.

There is a related core issue at [Allow updating modules with new service dependencies](https://www.drupal.org/node/2863986).

### LINKED ISSUE #4216: `config:import` should use the `UpdateKernel`
**Describe the bug**
I am trying to deploy a new feature that introduces a new custom module in my project. In this module there is a new service which will be used by some of the existing services in the existing modules.

So in the production database this new module with the new service dependency is not yet enabled. The new module will be enabled when I run `drush config:import`, but unfortunately when I run this command I get the following error:

```
    [drush] Executing command: ./vendor/bin/drush --root="./web" --verbose --yes config:import

In CheckExceptionOnInvalidReferenceBehaviorPass.php line 31:

  [Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException]
  The service "asset_release.fulfillment_guard" has a dependency on a non-existent service "workflow_state_permission.workflow_state_permission".
```

This is because the dependency container is built using the standard `DrupalKernel` which detects if a non-existing dependency is used and refuses to build. We should instead use the `UpdateKernel` which will skip all unmet dependencies automatically (ref. `UpdateCompilerPass::process()`). This has been introduced recently in Drupal 8.7.x, ref [Change record #3067480: The UpdateKernel removes services with unmet dependencies from the container](https://www.drupal.org/node/3067480).

I naively tried to add the annotation `@kernel update` to the `ConfigImportCommands::import()` command but this doesn't work. The `config:import` command is itself a service, so it is only discovered _after_ Drupal is already fully bootstrapped in `\Drush\Application::bootstrapAndFind()`. Unfortunately this will bootstrap the default `DrupalKernel` and will crash before it even has the chance to discover the command.

**To Reproduce**
* Add a new module that introduces a new service dependency.
* Run `drush config:import`

**Expected behavior**
* The config is imported.

**Actual behavior**
* `Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: The service "existing_module.existing_service" has a dependency on a non-existent service "new_module.new_service".`

**Workaround**
* Not tested but probably it is possible to enable the new module in a `hook_update_n()` hook before running `drush config:import`.

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 9.x
| Drupal version? | 8.7.x\|8.8.x
| PHP version     | 5.x\|7.x
| OS?             | Mac\|Linux\|Windows\|Unix

### PR DESCRIPTION
**Pull request by pfrenssen**
Fixes #4216

edit: Fixes #3574

### CODE REVIEW

**Comment by weitzman** [comment]
This looks good to me. Perhaps @greg-1-anderson has thoughts on whether this is the right pattern (registering a drupal service and calling it from a pre-bootstrap command command). Maybe it would be more better/explicit if the \Drush\Drupal\Commands\config\ConfigImportCommands::import were a hidden Drush command that was called by \Drush\Commands\config\ConfigImportCommands::import.

Also see #3574. Is that fixed or influenced by this PR?

**Comment by pfrenssen** [comment]
I can confirm that #3574 is exactly the same issue. I will close mine and add the info there.

**Comment by pfrenssen** [comment]
The test is green for all environments except "PHP 7.2 - Highest". `Unish\ConfigCase::testConfigExportImportStatusExistingConfig` is failing.

**Comment by weitzman** [comment]
Thats a known issue, unrelated to the PR.

**Comment by weitzman** [comment]
So UpdateKernel isnt really a reference to update.php? Its useful for config:import as well?

**Comment by pfrenssen** [comment]
The `UpdateKernel` has some adaptations that allow it to bootstrap Drupal when it is in the middle of an update. A deployment happens in the following stages:

1. New code is deployed.
2. First round of database updates - `hook_update_n()`.
3. Import configuration
4. Perform post-config-import tasks (will be added in https://www.drupal.org/project/drupal/issues/2901418)
5. Clear the cache (depending on the project, either all of them or only specific bins).

During this procedure the site is considered to be in a "broken state" because there might be discrepancies between the code, database and cache that can cause fatal errors. At the end of these 5 steps everything is in sync again and the site is back in a working state. 

Here is how the `UpdateKernel` protects against fatal errors during the update:

-  It prevents services critical to the deployment process from using the cache (`\Drupal\Core\Update\UpdateServiceProvider::alter()`).
- It forces a container rebuild (`\Drupal\Core\Update\UpdateKernel::initializeContainer()`).
- It clears broken services from the dependency injection container (`\Drupal\Core\Update\UpdateCompilerPass::process()`).
- It prevents the modified container from being cached so that it will be regenerated in correct state after the update is finished (`\Drupal\Core\Update\UpdateKernel::cacheDrupalContainer()`).

This PR is fixing step 3 in the update procedure. At this point any new modules that are added during the deployment are not enabled yet (they will soon be enabled when the `core.extension` config item is updated) so the site is still in a broken state. More specifically, the container cannot be compiled if any previously existing service depends on a service from one of the not-yet-enabled modules.

So we need to use the `UpdateKernel` in this step.

**Comment by pfrenssen** [comment]
I have in the meantime tried this solution in a test deployment of a branch that was exhibiting the problem in our "large" project. It now works perfectly.

File: src/Drupal/Commands/config/ConfigImportCommands.php [hunk 1]
```diff
@@ -177,14 +177,10 @@ public function __construct(ConfigManagerInterface $configManager, StorageInterf
     /**
      * Import config from a config directory.
      *
-     * @command config:import
```
Comment (weitzman) [inline src/Drupal/Commands/config/ConfigImportCommands.php]: Can you see if the validate() hook in this file still runs? Even if it does, I think it should move to the new file along with its command definition.

Comment (pfrenssen) [reply src/Drupal/Commands/config/ConfigImportCommands.php]: It is still working:

```
$ ./vendor/bin/drush config:import --partial
 [error]  Enable the config module in order to use the --partial option.
```

I'll move it to the new file anyway.

**Comment by claudiu-cristea** [comment]
I've provided a regression test in #4220. It's in a separate PR because I wanted to see the failure against master

**Comment by claudiu-cristea** [comment]
Now the PR #4220 is against config-import-updatekernel branch. If is accepted will be merged here.

**Comment by weitzman** [comment]
Thanks much. I feel this is a bit risky to backport to 9, but am happy to be convinced otherwise.

**Comment by pfrenssen** [comment]
Thanks!

There is no pressing need for us to have this backported to Drush 9. The patch from this PR applies cleanly to 9 and can be used for now by us and others who are hit with this problem. And Drush 10 is just around the corner!

**Comment by alexpott** [comment]
This needs to be rolled back. It does work for the most basic of configuration imports that enable a module. Fundamentally the config import operation is not an update operation.

Steps to reproduce:
1. Install standard.
2. Install media_library
3. export configuration
4. visit admin/config/media/media-library
5. uninstall media_library
6. use drush to import configuration
7. visit admin/config/media/media-library - it's broken
8. use drush cr but it is still broken but the module is showing as enabled and there's nothing to import.

This is happening because of https://www.drupal.org/node/3055443 but that doesn't make this change correct. The original issue behind this feels like it is the wrong way around. Delivering config updates and code updates shouldn't happen at the same time. If your updating modules to depend on a new service provided by a new module then you need to ship a code update that enables that module via hook_update_N. 

This change has made Drush's config import work in a different way that cores that is not supported. If we want to support that - and I'm not sure we do - then we need to change core's config ui as well and do this in a co-ordinated fashion.

**Comment by BramDriesen** [comment]
Cross referencing a bit here: https://www.drupal.org/project/drupal/issues/3119373

File: tests/functional/ConfigTest.php [hunk 2]
```diff
@@ -83,4 +83,52 @@ public function testConfigExportImportStatusExistingConfig()
...
+        $extensionFile = $this->getConfigSyncDir() . '/core.extension.yml';
+        $this->assertFileExists($extensionFile);
+        $extension = Yaml::decode(file_get_contents($extensionFile));
+        $extension['module']['devel'] = 0;
+        require_once $root . "/core/includes/module.inc";
+        $extension['module'] = module_config_sort($extension['module']);
+        file_put_contents($extensionFile, Yaml::encode($extension));
+
+        $this->drush('config:import');
+    }
```
Comment (alexpott) [inline tests/functional/ConfigTest.php]: This is not actually testing that config import works. It's making an assumption that if the command is successful it's successful but it doesn't actually properly test that a module has been enabled and it in the container.modules parameter.

**Comment by BramDriesen** [comment]
@weitzman  Please have a look at this (not sure if you notified on comments of closed PR's)

**Comment by weitzman** [comment]
Thanks Bram. We are awaiting a reply from @claudiu-cristea  and @pfrenssen about this. Please join #drush on Slack as there may be conversation there as well.

**Comment by pfrenssen** [comment]
> Delivering config updates and code updates shouldn't happen at the same time. If your updating modules to depend on a new service provided by a new module then you need to ship a code update that enables that module via hook_update_N.

I'm not sure how to handle this situation when using automated deployments. Isn't it the case that in normal circumstances the config import happens after the db updates but before the post updates? For example a typical post-update script will handle a data migration to a new field which is deployed through config. How are we supposed to enable a module in a post-update before deploying config without resorting to a manual deployment?

There is also the problem that you don't have real granular control over when and how users deploy your project, they might skip a bunch of releases and go e.g. from version 1.10 to 1.20 in one deployment. How are we supposed to deal with this if there is a post-update hook in version 1.12 that enables a module and should be run before any of the config from later versions is deployed?

**Comment by BramDriesen** [comment]
@pfrenssen even just enabling a contributed/core module (that might even be on the server already for a long time) does not work...

**Comment by pfrenssen** [comment]
@BramDriesen can you open an issue for this providing more information about the specific circumstances that you encounter?

This is definitely not a generic case, looking back in our project history we have enabled 21 modules and disabled 2 modules using this code since it was released and had not a single problem.

**Comment by BramDriesen** [comment]
@pfrenssen see #4350 referenced above.

**Comment by bartlangelaan** [comment]
I can confirm that I experienced this issue, even when only enabling modules. Applying the patch from issue [#3119373](https://www.drupal.org/project/drupal/issues/3119373#comment-13515436) fixed this problem.

I executed this command, once with patch and once without patch. 

```
ddev import-db < ../qfyruyad6alua--master-7rqtwti--drupal-database--main--dump.sql && \
ddev exec drush -y cache-rebuild && \
ddev exec drush -y updatedb && \
ddev exec drush -y config-import --diff && \
ddev exec drush -y cache-rebuild && \
curl https://[project]/graphql
```

In the end of the command result, you can see that without that patch the `graphql` module does not get properly installed, because the 403 path is loaded. With the patch, 403 is loaded (the path exists, just not the right permissions). This is expected behaviour.


<details><summary>Without patch</summary>

```
Successfully imported database 'db' for [project] 
Existing database 'db' was dropped before importing  
```

```
 [success] Cache rebuild complete.
```

```
 [success] No pending updates.
```

``` 
diff --git a/tmp/drush_tmp_1585601413_5e825b85b68a7/core.extension.yml b/tmp/drush_tmp_1585601413_5e825b85c6e83/core.extension.yml
index 2403cd61d..87d84b10c 100644
--- a/tmp/drush_tmp_1585601413_5e825b85b68a7/core.extension.yml
+++ b/tmp/drush_tmp_1585601413_5e825b85c6e83/core.extension.yml
@@ -12,6 +12,9 @@ module:
   field_ui: 0
   file: 0
   filter: 0
+  graphql: 0
+  graphql_core: 0
+  graphql_export: 0
   link: 0
   menu_ui: 0
   node: 0


 // Import the listed configuration changes?: yes.                                                                      

 [notice] Synchronized extensions: install graphql.
 [notice] Synchronized extensions: install graphql_core.
 [notice] Synchronized extensions: install graphql_export.
 [notice] Finalizing configuration synchronization.
 [success] The configuration was imported successfully.
```

``` 
 [success] Cache rebuild complete.
```

```
system.404
```
</details>

<details><summary>With patch</summary>

```
Successfully imported database 'db' for [project] 
Existing database 'db' was dropped before importing
```

``` 
 [success] Cache rebuild complete.
```

```
 [success] No pending updates.
```

```
diff --git a/tmp/drush_tmp_1585601997_5e825dcdde7e6/core.extension.yml b/tmp/drush_tmp_1585601997_5e825dcdf2c8b/core.extension.yml
index 2403cd61d..87d84b10c 100644
--- a/tmp/drush_tmp_1585601997_5e825dcdde7e6/core.extension.yml
+++ b/tmp/drush_tmp_1585601997_5e825dcdf2c8b/core.extension.yml
@@ -12,6 +12,9 @@ module:
   field_ui: 0
   file: 0
   filter: 0
+  graphql: 0
+  graphql_core: 0
+  graphql_export: 0
   link: 0
   menu_ui: 0
   node: 0


 // Import the listed configuration changes?: yes.                                                                      

 [notice] Synchronized extensions: install graphql.
 [notice] Synchronized extensions: install graphql_core.
 [notice] Synchronized extensions: install graphql_export.
 [notice] Finalizing configuration synchronization.
 [success] The configuration was imported successfully.
```


```
 [success] Cache rebuild complete.
```

```
system.403
```
</details>

So, the patch fixed it.

But when I applied the [Drush patch](https://www.drupal.org/project/drupal/issues/3119373#comment-13510850) from the same Drupal issue, the same (expected) result appeared.

So, it looks like both patches fix this problem.

Maybe the update kernel should only be used from 8.8.5 onwards, if at all?

**Pull request merged**